File: backend/src/services/websocket.service.js
import { Server } from "socket.io";
import jwt from "jsonwebtoken";

class WebSocketService {
  constructor() {
    this.io = null;
    this.activeRooms = new Map(); // roomId -> Set of user data
  }

  /**
   * Initialize WebSocket server
   * @param {Server} httpServer - HTTP server instance
   */
  initialize(httpServer) {
    this.io = new Server(httpServer, {
      cors: {
        origin: (origin, callback) => {
          // Allow all localhost origins in development
          if (!origin || /^http:\/\/localhost:\d+$/.test(origin)) {
            return callback(null, true);
          }
          callback(new Error("Not allowed by CORS"));
        },
        credentials: true,
      },
    });

    this.setupConnectionHandlers();
    console.log("✅ WebSocket service initialized");
  }

  /**
   * Setup connection handlers
   */
  setupConnectionHandlers() {
    // Middleware for authentication
    this.io.use((socket, next) => {
      const token =
        socket.handshake.auth.token ||
        socket.handshake.headers.authorization?.split(" ")[1];
      if (!token) {
        // Allow unauthenticated connections for now if needed, or reject
        // For now, let's just log and proceed, but ideally we should reject or handle guest users
        // But for notifications, we need userId.
        // If no token, maybe it's a guest or public page?
        // Let's try to verify if token exists.
        return next();
      }
      try {
        const decoded = jwt.verify(token, process.env.JWT_SECRET);
        socket.userId = decoded.id;
        next();
      } catch (err) {
        console.error("Socket auth error:", err.message);
        next(new Error("Authentication error"));
      }
    });

    this.io.on("connection", (socket) => {
      console.log(`🔌 Client connected: ${socket.id}, User: ${socket.userId}`);

      // Join user specific room
      if (socket.userId) {
        socket.join(`user_${socket.userId}`);
      }

      // Join a collaboration room (task, project, etc.)
      socket.on("join_room", ({ roomId, userId, userName }) => {
        socket.join(roomId);

        // Track active users in room
        if (!this.activeRooms.has(roomId)) {
          this.activeRooms.set(roomId, new Map());
        }
        this.activeRooms.get(roomId).set(socket.id, { userId, userName });

        // Notify others in room
        const activeUsers = Array.from(this.activeRooms.get(roomId).values());
        this.io.to(roomId).emit("room_users_updated", { activeUsers });

        console.log(`👥 User ${userName} joined room ${roomId}`);
      });

      // Leave a collaboration room
      socket.on("leave_room", ({ roomId }) => {
        this.leaveRoom(socket, roomId);
      });

      // Cursor position update
      socket.on("cursor_update", ({ roomId, userId, userName, position }) => {
        socket.to(roomId).emit("cursor_moved", {
          userId,
          userName,
          position,
          socketId: socket.id,
        });
      });

      // Content update (description, comment, etc.)
      socket.on(
        "content_update",
        ({ roomId, contentType, contentId, content, userId, userName }) => {
          socket.to(roomId).emit("content_changed", {
            contentType,
            contentId,
            content,
            userId,
            userName,
            timestamp: new Date().toISOString(),
          });
        }
      );

      // Comment added
      socket.on("comment_added", ({ roomId, comment, userId, userName }) => {
        this.io.to(roomId).emit("new_comment", {
          comment,
          userId,
          userName,
          timestamp: new Date().toISOString(),
        });
      });

      // Notification event
      socket.on("notification_received", ({ userId, notification }) => {
        this.io.to(`user_${userId}`).emit("new_notification", {
          notification,
          timestamp: new Date().toISOString(),
        });
      });

      // Typing indicator
      socket.on("typing_start", ({ roomId, userId, userName, field }) => {
        socket.to(roomId).emit("user_typing", {
          userId,
          userName,
          field,
          isTyping: true,
        });
      });

      socket.on("typing_stop", ({ roomId, userId, userName, field }) => {
        socket.to(roomId).emit("user_typing", {
          userId,
          userName,
          field,
          isTyping: false,
        });
      });

      // Handle disconnection
      socket.on("disconnect", () => {
        console.log(`🔌 Client disconnected: ${socket.id}`);
        this.handleDisconnection(socket);
      });
    });
  }

  /**
   * Handle user leaving a room
   */
  leaveRoom(socket, roomId) {
    socket.leave(roomId);

    if (this.activeRooms.has(roomId)) {
      this.activeRooms.get(roomId).delete(socket.id);

      // Notify remaining users
      const activeUsers = Array.from(this.activeRooms.get(roomId).values());
      this.io.to(roomId).emit("room_users_updated", { activeUsers });

      // Clean up empty rooms
      if (this.activeRooms.get(roomId).size === 0) {
        this.activeRooms.delete(roomId);
      }
    }
  }

  /**
   * Handle client disconnection
   */
  handleDisconnection(socket) {
    // Remove user from all rooms
    this.activeRooms.forEach((users, roomId) => {
      if (users.has(socket.id)) {
        this.leaveRoom(socket, roomId);
      }
    });
  }

  /**
   * Broadcast event to a specific room
   */
  broadcastToRoom(roomId, event, data) {
    if (this.io) {
      this.io.to(roomId).emit(event, data);
    }
  }

  /**
   * Emit event to a specific user
   */
  emitToUser(userId, event, data) {
    if (this.io) {
      this.io.to(`user_${userId}`).emit(event, data);
    }
  }

  /**
   * Get active users in a room
   */
  getActiveUsers(roomId) {
    if (!this.activeRooms.has(roomId)) {
      return [];
    }
    return Array.from(this.activeRooms.get(roomId).values());
  }
}

// Singleton instance
const websocketService = new WebSocketService();
export default websocketService;

File: creapolis_app/lib/core/services/websocket_service.dart
import 'package:socket_io_client/socket_io_client.dart' as io;
import 'package:logger/logger.dart';

/// Service for managing WebSocket connections for real-time collaboration
class WebSocketService {
  io.Socket? _socket;
  final Logger _logger = Logger();
  final String baseUrl;
  bool _isConnected = false;

  // Callbacks
  Function(List<Map<String, dynamic>>)? onRoomUsersUpdated;
  Function(Map<String, dynamic>)? onCursorMoved;
  Function(Map<String, dynamic>)? onContentChanged;
  Function(Map<String, dynamic>)? onNewComment;
  Function(Map<String, dynamic>)? onUserTyping;

  WebSocketService({required this.baseUrl});

  /// Check if WebSocket is connected
  bool get isConnected => _isConnected;

  /// Connect to WebSocket server
  Future<void> connect({required String token}) async {
    try {
      _logger.i('Connecting to WebSocket server: $baseUrl');

      _socket = io.io(
        baseUrl,
        io.OptionBuilder()
            .setTransports(['websocket'])
            .enableAutoConnect()
            .enableReconnection()
            .setReconnectionDelay(1000)
            .setReconnectionDelayMax(5000)
            .setReconnectionAttempts(5)
            .setExtraHeaders({'Authorization': 'Bearer $token'})
            .build(),
      );

      _setupEventHandlers();

      _socket?.connect();
    } catch (e) {
      _logger.e('Error connecting to WebSocket: $e');
    }
  }

  /// Setup event handlers
  void _setupEventHandlers() {
    _socket?.onConnect((_) {
      _isConnected = true;
      _logger.i('✅ WebSocket connected');
    });

    _socket?.onDisconnect((_) {
      _isConnected = false;
      _logger.w('🔌 WebSocket disconnected');
    });

    _socket?.onConnectError((error) {
      _logger.e('❌ WebSocket connection error: $error');
    });

    _socket?.onError((error) {
      _logger.e('❌ WebSocket error: $error');
    });

    // Room users updated
    _socket?.on('room_users_updated', (data) {
      _logger.d('👥 Room users updated: $data');
      if (onRoomUsersUpdated != null && data != null) {
        final activeUsers = (data['activeUsers'] as List)
            .map((user) => user as Map<String, dynamic>)
            .toList();
        onRoomUsersUpdated!(activeUsers);
      }
    });

    // Cursor moved
    _socket?.on('cursor_moved', (data) {
      _logger.d('🖱️ Cursor moved: $data');
      if (onCursorMoved != null && data != null) {
        onCursorMoved!(data as Map<String, dynamic>);
      }
    });

    // Content changed
    _socket?.on('content_changed', (data) {
      _logger.d('📝 Content changed: $data');
      if (onContentChanged != null && data != null) {
        onContentChanged!(data as Map<String, dynamic>);
      }
    });

    // New comment
    _socket?.on('new_comment', (data) {
      _logger.d('💬 New comment: $data');
      if (onNewComment != null && data != null) {
        onNewComment!(data as Map<String, dynamic>);
      }
    });

    // User typing
    _socket?.on('user_typing', (data) {
      _logger.d('⌨️ User typing: $data');
      if (onUserTyping != null && data != null) {
        onUserTyping!(data as Map<String, dynamic>);
      }
    });
  }

  /// Join a collaboration room
  void joinRoom({
    required String roomId,
    required int userId,
    required String userName,
  }) {
    if (!_isConnected) {
      _logger.w('Cannot join room: WebSocket not connected');
      return;
    }

    _logger.i('Joining room: $roomId as $userName');
    _socket?.emit('join_room', {
      'roomId': roomId,
      'userId': userId,
      'userName': userName,
    });
  }

  /// Leave a collaboration room
  void leaveRoom({required String roomId}) {
    if (!_isConnected) return;

    _logger.i('Leaving room: $roomId');
    _socket?.emit('leave_room', {'roomId': roomId});
  }

  /// Send cursor position update
  void updateCursor({
    required String roomId,
    required int userId,
    required String userName,
    required Map<String, dynamic> position,
  }) {
    if (!_isConnected) return;

    _socket?.emit('cursor_update', {
      'roomId': roomId,
      'userId': userId,
      'userName': userName,
      'position': position,
    });
  }

  /// Send content update
  void updateContent({
    required String roomId,
    required String contentType,
    required String contentId,
    required String content,
    required int userId,
    required String userName,
  }) {
    if (!_isConnected) return;

    _socket?.emit('content_update', {
      'roomId': roomId,
      'contentType': contentType,
      'contentId': contentId,
      'content': content,
      'userId': userId,
      'userName': userName,
    });
  }

  /// Add a comment
  void addComment({
    required String roomId,
    required Map<String, dynamic> comment,
    required int userId,
    required String userName,
  }) {
    if (!_isConnected) return;

    _socket?.emit('comment_added', {
      'roomId': roomId,
      'comment': comment,
      'userId': userId,
      'userName': userName,
    });
  }

  /// Start typing indicator
  void startTyping({
    required String roomId,
    required int userId,
    required String userName,
    required String field,
  }) {
    if (!_isConnected) return;

    _socket?.emit('typing_start', {
      'roomId': roomId,
      'userId': userId,
      'userName': userName,
      'field': field,
    });
  }

  /// Stop typing indicator
  void stopTyping({
    required String roomId,
    required int userId,
    required String userName,
    required String field,
  }) {
    if (!_isConnected) return;

    _socket?.emit('typing_stop', {
      'roomId': roomId,
      'userId': userId,
      'userName': userName,
      'field': field,
    });
  }

  /// Disconnect from WebSocket server
  void disconnect() {
    _logger.i('Disconnecting from WebSocket server');
    _socket?.disconnect();
    _socket?.dispose();
    _isConnected = false;
  }

  /// Dispose resources
  void dispose() {
    disconnect();
  }
}
